clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -5637,6 +5638,38 @@ static bool EvaluateUnaryTypeTrait(Sema &Self, TypeTrait UTT,
     return false;
   case UTT_IsTriviallyEqualityComparable:
     return isTriviallyEqualityComparableType(Self, T, KeyLoc);
+  case UTT_IsImplicitLifetime: {
+    DiagnoseVLAInCXXTypeTrait(Self, TInfo,
+                              tok::kw___builtin_is_implicit_lifetime);
+
+    // [basic.types.general] p9
+    // Scalar types, implicit-lifetime class types ([class.prop]),
+    // array types, and cv-qualified versions of these types
+    // are collectively called implicit-lifetime types.
+    QualType UnqualT = T->getCanonicalTypeUnqualified();
```

Comment:
How should `_Atomic int` behave? We support that as an extension in C++

Reply:
I don't think we want to old this PR on _Atomic, but I don't know what the answer ought to be.
Rejecting that code all together might be the most conservative solution.

@zygoloid

Reply:
After an offline discussion, I believe we agreed on not doing anything special for `_Atomic` for the lack of better alternatives. This can result in discrepancies between `__builtin_is_implicit_lifetime(std::atomic<T>)` (that might yield `true` depending on `std::atomic` implementation) and `__builtin_is_implicit_lifetime(_Atomic T)` (which is always `false`), but that's not new (https://godbolt.org/z/7e7954b8c).

Reply:
I think we should reject use of `_Atomic` so that we don't give a true or false answer; that leaves the door open for us to better define the behavior in the future without risk of breaking code.

Reply:
Done. We're rejecting atomic types now.